lib/FrontendTool/ReferenceDependencies.cpp

```diff
@@ -23,253 +23,308 @@
+    // Before writing to the dependencies file path, preserve any previous file
+    // that may have been there. No error handling -- this is just a nicety, it
+    // doesn't matter if it fails.
+    llvm::sys::fs::rename(outputPath, outputPath + "~");
+
+    std::error_code EC;
+    auto out = llvm::make_unique<llvm::raw_fd_ostream>(outputPath, EC,
+                                                       llvm::sys::fs::F_None);
+
+    if (out->has_error() || EC) {
+      diags.diagnose(SourceLoc(), diag::error_opening_output, outputPath,
+                     EC.message());
+      out->clear_error();
+      return None;
     }
+
+    return ReferenceDependencyFileEmitter(SF, depTracker, std::move(out));
   }
 
-  return VD->getFormalAccess() <= AccessLevel::FilePrivate;
-}
+  void emit() {
+    *out << "### Swift dependencies file v0 ###\n";
 
-static bool extendedTypeIsPrivate(TypeLoc inheritedType) {
-  auto type = inheritedType.getType();
-  if (!type)
-    return true;
+    emitProviders();
+    emitDependencies();
+    emitInterfaceHash();
+    }
 
-  if (!type->isExistentialType()) {
-    // Be conservative. We don't know how to deal with other extended types.
-    return false;
+    void emitProviders() const;
+    void emitDependencies() const;
+
+    void emitInterfaceHash() {
+      llvm::SmallString<32> interfaceHashVal;
+      SF->getInterfaceHash(interfaceHashVal);
+      *out << interfaceHash << ": \"" << interfaceHashVal << "\"\n";
   }
+};
 
-  auto layout = type->getExistentialLayout();
-  assert(!layout.superclass && "Should not have a subclass existential "
-         "in the inheritance clause of an extension");
-  for (auto protoTy : layout.getProtocols()) {
-    if (!declIsPrivate(protoTy->getDecl()))
-      return false;
+class ProviderEmitter {
+  llvm::MapVector<const NominalTypeDecl *, bool> extendedNominals;
+  llvm::SmallVector<const FuncDecl *, 8> memberOperatorDecls;
+  llvm::SmallVector<const ExtensionDecl *, 8> extensionsWithJustMembers;
+
+  llvm::raw_fd_ostream &out;
+
+public:
+  ProviderEmitter(llvm::raw_fd_ostream &out) : out(out) {}
+
+  void emit(const SourceFile *SF);
+
+private:
+  void emitTopLevelNames(const ArrayRef<const Decl *> decls);
+  void emitTopLevelDecl(const Decl *D);
+  void emitExtensionDecl(const ExtensionDecl *ED);
+  void emitNominalTypeDecl(const NominalTypeDecl *NTD);
+  void emitValueDecl(const ValueDecl *VD) const;
+
+  void emitNominalTypes() const;
+  void emitMembers() const;
+  void emitDynamicLookupMembers(const SourceFile *SF) const;
+
+  void findNominalsAndOperators(const DeclRange members);
+  static bool extendedTypeIsPrivate(const TypeLoc inheritedType);
+  static bool declIsPrivate(const Decl *member);
+
+  template <typename Decl_T> static bool hasVisibleName(const Decl_T *decl) {
+    return decl->hasName() && isVisibleOutsideItsFile(decl);
   }
+};
 
-  return true;
+void ReferenceDependencyFileEmitter::emitProviders() const {
```

Comment:
I'm not sure this function pulls its weight. It's an extra hop for anyone reading the code.

Reply:
I know, but it's an intention change; it maps the activity to the method. And I like the way its caller reads:
```
emitProviders();
emitDependencies();
emitInterfaceHash();
```